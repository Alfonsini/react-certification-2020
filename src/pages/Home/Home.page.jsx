import React, { useContext } from 'react';
import { useHistory } from 'react-router-dom';

import { StyledCol, StyledRow } from './Home.styles';

import Video from '../../components/Video';

import { VideoContext } from '../../providers/Videos';
import { useFavoriteVideos } from '../../utils/hooks';

function HomePage() {
  const { state, dispatch } = useContext(VideoContext);
  const history = useHistory();
  const { favoriteVideos, setFavoriteVideosValue } = useFavoriteVideos();

  const handleVideoClick = async (id) => {
    const video = state.videosList.items.filter((v) => v.id.videoId === id)[0];

    if (!video) return;

    dispatch({
      type: 'SELECT_VIDEO',
      payload: {
        currentVideoId: id,
        currentVideoTitle: video.snippet.title,
        currentVideoDescription: video.snippet.description,
        currentVideoIsFavorite: video.isFavorite,
      },
    });

    history.push('/video-details');
  };

  const handleChangeFavorite = (id) => {
    if (!id) return;

    const video = state.videosList.items.filter((v) => v.id.videoId === id);

    if (video && video.length > 0) {
      const index = state.videosList.items.indexOf(video[0]);
      state.videosList.items[index].isFavorite = !state.videosList.items[index]
        .isFavorite;

      dispatch({ type: 'SET_VIDEOS', payload: { videosList: state.videosList } });

      let videos = favoriteVideos;
      if (state.videosList.items[index].isFavorite) {
        videos.push(video[0]);
      } else {
        videos = favoriteVideos.filter((v) => v.id.videoId !== id);
      }

      setFavoriteVideosValue(videos);
    }
  };

  return (
    <StyledRow data-testid="home-page">
      {!state.isLoading &&
        state.videosList &&
        state.videosList.items &&
        state.videosList.items.map((v) => (
          <StyledCol key={v.id.videoId.toString()} data-testid="home-video-col">
            <Video
              id={v.id.videoId}
              title={v.snippet.title}
              channelId={v.snippet.channelId}
              channelName={v.snippet.channelTitle}
              image={v.snippet.thumbnails.medium.url}
              publishedAt={v.snippet.publishedAt}
              onVideoClick={handleVideoClick}
              isFavorite={v.isFavorite}
              onChangeFavorite={handleChangeFavorite}
            />
          </StyledCol>
        ))}
      {!state.isLoading && state.errorMessage && (
        <span data-testid="error-msg">{state.errorMessage}</span>
      )}
    </StyledRow>
  );
}

export default HomePage;
