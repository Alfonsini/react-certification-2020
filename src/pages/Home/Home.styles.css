.col {
  margin: 5px 5px 10px 5px;
  padding: 6px 6px 6px 6px;
}

.row {
  margin: 0;
  width: 95.5%;

  left: 0;
  right: 0;

  display: flex;
  justify-content: center;
  align-items: left;
  flex-flow: column;
  flex-direction: row;
  flex-wrap: wrap;
}
