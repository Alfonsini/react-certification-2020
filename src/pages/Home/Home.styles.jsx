import styled from 'styled-components';

const StyledCol = styled.div`
  margin: 5px 5px 10px 5px;
  padding: 6px 6px 6px 6px;
`;

const StyledRow = styled.div`
  margin: 0;
  width: 95.5%;

  left: 0;
  right: 0;

  display: flex;
  justify-content: center;
  align-items: left;
  flex-flow: column;
  flex-direction: row;
  flex-wrap: wrap;
`;

export { StyledCol, StyledRow };
