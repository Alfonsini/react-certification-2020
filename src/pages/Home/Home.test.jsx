import React from 'react';
import { render, screen } from '@testing-library/react';

import Home from '.';

beforeEach(() => {
  render(<Home />);
});

afterEach(() => {});

describe('Home', () => {
  test('home render correctly', async () => {
    const aside = screen.getByTestId('row');
    expect(aside).toMatchSnapshot();
  });
});
