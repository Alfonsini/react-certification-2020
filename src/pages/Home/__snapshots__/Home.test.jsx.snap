// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home home render correctly 1`] = `
<div
  class="sc-iBaPrD gTSUvL"
  data-testid="home-page"
>
  
</div>
`;
